drivers/flash/spi_flash.c

```diff
@@ -0,0 +1,523 @@
+			   rxbufs, ARRAY_SIZE(rxbufs));
+	k_sem_give(&data->sem);
+
+	return r;
+}
+
+static int spi_flash_write(struct device *dev, off_t offset,
+			      const void *ptr, size_t len)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	u8_t hdr[5];
+	struct spi_buf txbufs[] = {
+		{ .buf = hdr,  .len = 1 + data->addr_len },
+		{ .buf = NULL, .len = data->write_dummy },
+		{ .buf = (void*)ptr, .len = len },
+	};
+	int r, i;
+
+	SYS_LOG_DBG("write: @%x len %x, opcode %x (page size %d)",
+		offset, len,
+		data->write_opcode, data->page_size);
+
+	if (!data->write_opcode) {
+		return -ENOTSUP;
+	}
+
+	if (offset < 0 || offset+len >= data->flash_size || len > data->page_size) {
+		return -ENODEV;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	wait_for_flash_idle(dev);
+	r = spi_flash_cmd(dev, CMD_WRITE_ENABLE);
+	if (r != 0) {
+		goto err;
+	}
+
+	i = 0;
+	hdr[i++] = data->write_opcode;
+	if (data->addr_len == 4) {
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
+	hdr[i++] = offset >> 8;
+	hdr[i++] = offset;
+	r = spi_transceive(&data->config, txbufs, ARRAY_SIZE(txbufs), NULL, 0);
+	if (r == 0) {
+		wait_for_flash_idle(dev);
+	}
+	k_sem_give(&data->sem);
+
+err:
+	return r;
+}
+
+
+static int spi_flash_write_protection_set(struct device *dev, bool enable)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	int r;
+
+	k_sem_take(&data->sem, K_FOREVER);
+	wait_for_flash_idle(dev);
+	if (enable) {
+		r = spi_flash_cmd(dev, CMD_WRITE_DISABLE);
+	} else {
+		SYS_LOG_DBG("global unblock %02x", data->global_unprotect_opcode);
+
+		r = spi_flash_cmd(dev, CMD_WRITE_ENABLE);
+		if (r == 0 && data->global_unprotect_opcode) {
+			wait_for_flash_idle(dev);
+			r = spi_flash_cmd(dev, data->global_unprotect_opcode);
+		}
+	}
+	k_sem_give(&data->sem);
+
+	return r;
+}
+
+static inline int spi_flash_erase_cmd(struct device *dev, u8_t opcode, off_t offset)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	u8_t hdr[5];
+	struct spi_buf txbufs[] = {
+		{ .buf = hdr,  .len = 1 + data->addr_len },
+	};
+	int r, i;
+
+	SYS_LOG_DBG("erase: %02x %06x", opcode, offset);
+
+	wait_for_flash_idle(dev);
+	r = spi_flash_cmd(dev, CMD_WRITE_ENABLE);
+	if (r != 0) {
+		return r;
+	}
+
+	i = 0;
+	hdr[i++] = opcode;
+	if (data->addr_len == 4) {
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
+	hdr[i++] = offset >> 8;
+	hdr[i++] = offset;
+
+	wait_for_flash_idle(dev);
+	return spi_transceive(&data->config, txbufs, ARRAY_SIZE(txbufs), 0, 0);
+}
+
+static int spi_flash_erase(struct device *dev, off_t offset, size_t size)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	off_t offs = offset;
+	size_t left = size;
+	int ret = 0;
+
+	if (offset < 0 || offset+size >= data->flash_size ||
+	    (offset & 0xfff) || (size & 0xfff)) {
+		return -ENODEV;
+	}
+	if (!data->bfp.opcode_erase_4k) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	/* FIXME: Support chip erase, and large block erase */
```

Comment:
How are you sure all chips support 4k erase ?

Reply:
I think it's sort of requirement. In any case the presence of this opcode is tested above.